### PR TITLE
Type annotations for sqlalchemy.orm.dynamic

### PR DESCRIPTION
**Pull request by maxamly**
### Description

An attempt to annotate lib/sqlalchemy/orm/dynamic.py with type hints (issue https://github.com/sqlalchemy/sqlalchemy/issues/6810)

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

### CODE REVIEW

**Comment by zzzeek** [comment]
hi !

thanks for this!  can you remove the "ignores" that were added to files that otherwise weren't changed like asyncio/base.py  etc?   these pass with current mypy, if that's changed we have to fix that separately.

**Comment by mlatysh** [comment]
Hi @zzzeek !

I've made the changes you requested, but as you mentioned, there are still some mypy issues (attached). Do you mean that I should fix those separately in another pull request?
Снимок экрана 2022-12-30 в 10 21 46

**Comment by zzzeek** [comment]
hi -

what happens if you run a clean mypy 0.991 on the source tree without your changes?   Note that this code is all run through mypy 0.991 on an automated basis and there are currently no errors generated.      I see this line in asyncio is being highlighted by pylance but we are targeting mypy for now.

if OTOH the changes you make in dynamic somehow introduce mypy suddenly having issues with asyncio/base.py, that would be unusual as dynamic.py does not reference the asyncio extension.

**Comment by mlatysh** [comment]
Mypy run on source tree without applying proposed here changes goes without errors, but I assume removing the line here 
https://github.com/sqlalchemy/sqlalchemy/blob/f74de7bc0b886b3585536e2a3c7f1828ec8be083/lib/sqlalchemy/orm/dynamic.py#L7

makes it check all the modules it depends on. I consider that because removing this line causes all the errors above.

**Comment by zzzeek** [comment]
hmm not reproducing that, if i remove that single line from dynamic.py in a clean source tree, I get 26 or 27 errors in dynamic.py and no other files (command run: `mypy lib`)

**Comment by zzzeek** [comment]
I also recommend using `dmypy` for development, it's much faster

**Comment by mlatysh** [comment]
Sorry, my fault. I got some mypy errors on clean project's source tree, `main` brunch of repo. I've attached screenshot showing that (and mypy version). Sorry again for misunderstanding.
Снимок экрана 2023-01-02 в 12 56 55

**Comment by zzzeek** [comment]
ah.   OK upgrade your python to at least python 3.10 or python 3.11, you are on an old python version

**Comment by mlatysh** [comment]
I got you. I've fixed all mypy errors and now there are no mypy issues in proposed changes and changes affect only `dynamic.py` module

**Comment by zzzeek** [comment]
woop!  thanks!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 421977686f17ce1f8fa22f36e39c29f738ed93bb of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 421977686f17ce1f8fa22f36e39c29f738ed93bb: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

**Comment by CaselIT** [comment]
hi @mlatysh sorry for the delay on this.

it seems that the type checking is failing with the errors:
```
pep484 installed: greenlet==2.0.2,mypy==1.1.1,mypy-extensions==1.0.0,SQLAlchemy @ file:///home/jenkins/workspace/pep484/sqlalchemy/sqlalchemy/.tox/.tmp/package/1/SQLAlchemy-2.0.7.dev0.zip,typing_extensions==4.5.0
pep484 run-test-pre: PYTHONHASHSEED='423367632'
pep484 run-test: commands[0] | mypy ./lib/sqlalchemy
lib/sqlalchemy/orm/dynamic.py:72: error: Too few arguments  [call-arg]
lib/sqlalchemy/orm/dynamic.py:72: note: "autoflush" is considered instance variable, to make it class variable use ClassVar[...]
lib/sqlalchemy/orm/dynamic.py:73: error: Argument 1 has incompatible type "bool"; expected "_T"  [arg-type]
lib/sqlalchemy/orm/dynamic.py:104: error: Argument 4 to "__init__" of "AttributeImpl" has incompatible type "type"; expected "_LoaderCallable"  [arg-type]
Found 3 errors in 1 file (checked 246 source files)
```

**Comment by mlatysh** [comment]
Hello @CaselIT. I have made some changes to PR. The old version failed because of the new version of mypy (0.991 vs 1.1.1), now it should be fine. Please review my changes and let me know if you have any remarks or suggestions.

**Comment by CaselIT** [comment]
thanks. let's re-run

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 6b3a419f9d17bcd467ce8d18ed333c478ff42686 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset 6b3a419f9d17bcd467ce8d18ed333c478ff42686 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

**Comment by mlatysh** [comment]
Hello @CaselIT, I have made the changes as per your comments. Could you please rerun the tests when you have the time, and let me know if you have any further comments?

**Comment by CaselIT** [comment]
sure, thanks!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision dae9138f8019bf7e2b305ec869734373a997350d of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset dae9138f8019bf7e2b305ec869734373a997350d added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

File: lib/sqlalchemy/orm/dynamic.py [hunk 1]
```diff
@@ -102,18 +129,24 @@ class AppenderMixin(AbstractCollectionWriter[_T]):
 
     """
 
-    query_class = None
+    query_class: Type[Query[_T]] = None  # type: ignore[assignment]
```
Comment (sqla-tester) [inline lib/sqlalchemy/orm/dynamic.py:132]: **Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

File: lib/sqlalchemy/orm/dynamic.py [hunk 2]
```diff
@@ -270,7 +306,7 @@ class AppenderQuery(AppenderMixin[_T], Query[_T]):
     """
 
 
-def mixin_user_query(cls):
+def mixin_user_query(cls: Any) -> type[AppenderMixin[_T]]:
```
Comment (sqla-tester) [inline lib/sqlalchemy/orm/dynamic.py:309]: **Federico Caselli** (CaselIT) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

I've updated this change mike. Can you take a look?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

it's great can we also do writeonly.py and remove the mypy: ignore-untyped ?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/writeonly.py (line 127): why dont we do the whole file here?

**Comment by CaselIT** [comment]
Yes, it probably makes sense to just do both

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

updated

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/writeonly.py (line 127): Done

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

just one Q and then we can probably merge

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/attributes.py (line 838): too bad _CollectionFactoryType can't be a sub-type of _LoaderCallable, can it?

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/attributes.py (line 838): actually this is weird, the two callables have different interfaces.  this seems wrong

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/attributes.py (line 1122): yeah this is weird, why

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

replied. let me know if you want to go down that route

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/attributes.py (line 838): does not seem like it can, at least not without having a callable with any number of arg that does not make much sense.

to clean this up the best option would probably be to have an abstract AttributeImpl that does not have `callable_` and the `_fire_loader_callables` implementation (and all other stuff that's specific for the plain attribute impl)

Then the other impl can subclass that one instead of the plain attribute

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/attributes.py (line 1122): the issue is that collection attribute should not be a subclass of a plain attribute, but of an abstract superclass.

this method works only on a plain attributes, and is overloaded by the other ones

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

code review left on gerrit

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/attributes.py (line 1122): right the bug is that WriteOnly is putting the "typecallable" into this slot, that's not what this is.  fixing now.   writeonly's dont use the "typecallable", which is the callable that generates an empty collection.

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

now it's done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

* lib/sqlalchemy/orm/attributes.py (line 838): so this was an example of adding typing revealing a bug in the code.  that "typecallable" did not belong there
* lib/sqlalchemy/orm/attributes.py (line 1122): Done

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

great! mypy it's not happy though

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

**Comment by sqla-tester** [comment]
**Federico Caselli** (CaselIT) wrote:

you forgot an optional

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4507 has been **merged**. Congratulations! :)

**Pull request closed without merging**